src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,150 @@
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+
+        // Determine the equivalent amount of shares and burn them.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shares);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    /// @notice Redeem a specific amount of shares for underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param shareAmount The amount of shares to redeem for underlying tokens.
+    function redeem(address to, uint256 shareAmount) external virtual returns (uint256 underlyingAmount) {
+
+        // Determine the equivalent amount of underlying tokens.
+        underlyingAmount = shareAmount.fmul(exchangeRate(), baseUnit);
+
+        // Burn the provided amount of shares.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shareAmount);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    function beforeWithdraw(uint256 underlyingAmount) internal virtual {}
+
+    function afterDeposit(uint256 underlyingAmount) internal virtual {}
+
+    /*///////////////////////////////////////////////////////////////
+                        VAULT ACCOUNTING LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Returns a user's Vault balance in underlying tokens.
+    /// @param user The user to get the underlying balance of.
+    /// @return The user's Vault balance in underlying tokens.
+    function balanceOfUnderlying(address user) external view returns (uint256) {
+        return balanceOf[user].fmul(exchangeRate(), baseUnit);
+    }
+
+    /// @notice Returns the amount of underlying tokens a share can be redeemed for.
+    /// @return The amount of underlying tokens a share can be redeemed for.
+    function exchangeRate() public view returns (uint256) {
+        // Get the total supply of shares.
+        uint256 shareSupply = totalSupply;
+
+        // If there are no shares in circulation, return an exchange rate of 1:1.
+        if (shareSupply == 0) return baseUnit;
+
+        // Calculate the exchange rate by dividing the total holdings by the share supply.
+        return totalHoldings().fdiv(shareSupply, baseUnit);
+    }
+
+    /// @notice Calculates the total amount of underlying tokens the Vault holds.
+    /// @return totalUnderlyingHeld The total amount of underlying tokens the Vault holds.
+    function totalHoldings() public view virtual returns (uint256) {
+        return underlying.balanceOf(address(this));
```

Comment:
There are no checks for balances on the strategy contracts? Since on deposits you pass the underlying to another contract, totalHoldings will return much less than the actual holdings. The exchange rate will be 0 if there’s no underlying left on the vault contract and is fully deployed on strategy contracts.

Reply:
there are no strategy contracts in this demo implementation, its meant to be overriden